// users/install.js


// Default bootstrap script for moda Users

module.exports = function ( moda, cela ) {


  // Default Admin User

  moda.find( 'super' , 1 , function ( err, users ) {

    if ( users.length != 0 ) {

      return

    }

    moda.announce( 'Creating your default Admin account...'.cyan );

    var default_password = 'admin',

      user_object = {

        username: 'sysa',

        name: 'Cela Sysa',

        email: 'build@objas.com',

        roles: [ 'sysa', 'sysa-admin' ],

        super: 1,

        password: default_password

      };

    moda.user_new_save( user_object, function user_new_callback ( err, result ) {

      var user = result[0];

      moda.info(

        'Your admin username: '.cyan + user.username.yellow,
        'Your admin email: '.cyan + user.email.yellow,
        'Your admin password: '.cyan + default_password.yellow

        // NOTE // AN EMAIL WILL BE SENT TO build@objas.com TO SET PASSWORD

        // TODO // INSTEAD OF THIS AUTO ADMIN, A PROMPT OR GUI WILL BE USED

                // // AND USER WILL ADD THEIR OWN EMAIL

      );

      cela.sign( 'Cela' );

    });

  });


};
